### PR TITLE
lookup stores by thread id in addition to topic-partition

### PR DESCRIPTION
**Pull request by rodesai**

### CODE REVIEW

File: kafka-client/src/main/java/dev/responsive/kafka/internal/clients/ResponsiveKafkaClientSupplier.java [hunk 1]
```diff
@@ -226,6 +227,20 @@ private String threadIdFromProducerConfig(final String clientId) {
     return match.group(1);
   }
 
+  /**
+   * @param clientId the restore consumer client id
+   * @return the extracted StreamThread id, of the form "StreamThread-n"
+   */
+  private String threadIdFromRestoreConsumerConfig(final String clientId) {
```
Comment (agavra) [inline kafka-client/src/main/java/dev/responsive/kafka/internal/clients/ResponsiveKafkaClientSupplier.java:234]: why not just use the entire `clientId`?

File: kafka-client/src/main/java/dev/responsive/kafka/internal/clients/StoreCommitListener.java [hunk 2]
```diff
@@ -27,14 +28,14 @@ private void onCommit(
     for (final var e : committedOffsets.entrySet()) {
       final TopicPartition p = e.getKey().getPartition();
       for (final ResponsiveStoreRegistration storeRegistration
-          : storeRegistry.getRegisteredStoresForChangelog(p)) {
+          : storeRegistry.getRegisteredStoresForChangelog(p, threadId)) {
```
Comment (agavra) [inline kafka-client/src/main/java/dev/responsive/kafka/internal/clients/StoreCommitListener.java:31]: unrelated, but do you mind adding a comment about why we do this twice (once for committed offsets and once for written offsets)? it's not obviously clear to me what's the difference

File: kafka-client/src/main/java/dev/responsive/kafka/internal/utils/StoreUtil.java [hunk 3]
```diff
@@ -92,6 +93,17 @@ public static long durationToMillis(final Duration duration, final String parame
     }
   }
 
+  public static String streamThreadId() {
+    final String threadId = Thread.currentThread().getName();
+    final var regex = Pattern.compile(".*-(StreamThread-\\d+)");
```
Comment (agavra) [inline kafka-client/src/main/java/dev/responsive/kafka/internal/utils/StoreUtil.java:98]: ditto as above, why not just use the entirety of the thread name?

**Review by agavra — APPROVED** [review]
Awesome sleuthing @rodesai! Thanks :)

File: kafka-client/src/main/java/dev/responsive/kafka/internal/utils/StoreUtil.java [hunk 4]
```diff
@@ -92,6 +93,17 @@ public static long durationToMillis(final Duration duration, final String parame
     }
   }
 
+  public static String streamThreadId() {
```
Comment (agavra) [inline kafka-client/src/main/java/dev/responsive/kafka/internal/utils/StoreUtil.java:96]: Looks like we already have `ResponsiveMetrics#computeThreadId` - may want to consider reusing that (it also addresses Global store threads)

File: kafka-client/src/main/java/dev/responsive/kafka/internal/clients/ResponsiveKafkaClientSupplier.java (hunk 1 above)
Comment (rodesai) [reply kafka-client/src/main/java/dev/responsive/kafka/internal/clients/ResponsiveKafkaClientSupplier.java:234]: We need to be able to look up the store from all the different clients, so thread id is the finest granularity that's common to them all.

File: kafka-client/src/main/java/dev/responsive/kafka/internal/clients/StoreCommitListener.java (hunk 2 above)
Comment (rodesai) [reply kafka-client/src/main/java/dev/responsive/kafka/internal/clients/StoreCommitListener.java:31]: one path gets triggered for changelogs, the other for source topics (when the source table optimization is on)

File: kafka-client/src/main/java/dev/responsive/kafka/internal/stores/ResponsiveStoreRegistry.java [hunk 5]
```diff
@@ -41,10 +43,12 @@ public synchronized OptionalLong getCommittedOffset(final TopicPartition topicPa
   }
 
   public synchronized List<ResponsiveStoreRegistration> getRegisteredStoresForChangelog(
-      final TopicPartition topicPartition
+      final TopicPartition topicPartition,
+      final String threadId
   ) {
     return stores.stream()
-        .filter(s -> s.changelogTopicPartition().equals(topicPartition))
+        .filter(s -> s.changelogTopicPartition().equals(topicPartition)
```
[the remark this answers is not in the document]
Comment (rodesai) [reply kafka-client/src/main/java/dev/responsive/kafka/internal/stores/ResponsiveStoreRegistry.java:50]: no - I'm going to add a couple checks like this to this patch to tighten things up a bit. The other thing I'll check is that the offset passed to flush is greater than the last offset the store instance was aware of

File: kafka-client/src/main/java/dev/responsive/kafka/internal/utils/StoreUtil.java (hunk 3 above)
Comment (rodesai) [reply kafka-client/src/main/java/dev/responsive/kafka/internal/utils/StoreUtil.java:98]: yeah I guess we could. It's a bit more of a change and it was late and I was tired :)

**Pull request merged**